File: frontend/src/domain/ir/linkRules.ts
import type { LinkType, NodeKind } from '@/types';

export const LINK_RULES: Record<LinkType, Array<[NodeKind, NodeKind]>> = {
  realizes: [
    ['capability', 'goal'],
    ['task', 'capability'],
    ['flow', 'task'],
  ],
  supports: [
    ['task', 'capability'],
    ['flow_step', 'task'],
  ],
  performed_by: [
    ['task', 'actor'],
    ['flow_step', 'actor'],
  ],
  owns: [['actor', 'business_object']],
  precedes: [['flow_step', 'flow_step']],
  branches_to: [['flow_step', 'flow_step']],
  guards: [
    ['rule', 'flow_step'],
    ['rule', 'state_transition'],
  ],
  reads: [
    ['flow_step', 'business_object'],
    ['screen', 'business_object'],
    ['screen', 'actor'],
  ],
  writes: [
    ['flow_step', 'business_object'],
    ['screen', 'business_object'],
  ],
  changes_state: [
    ['flow_step', 'business_object'],
    ['state_transition', 'object_state'],
  ],
  depends_on: [
    ['capability', 'capability'],
    ['task', 'task'],
    ['task', 'capability'],
    ['flow_step', 'flow_step'],
    ['ui_component', 'flow_step'],
  ],
  contains: [
    ['goal', 'capability'],
    ['capability', 'capability'],
    ['business_object', 'field'],
    ['business_object', 'state_machine'],
    ['state_machine', 'object_state'],
    ['state_machine', 'state_transition'],
    ['screen', 'ui_component'],
    ['ui_component', 'ui_component'],
  ],
  accessible_by: [['screen', 'actor']],
  binds_field: [['ui_component', 'field']],
  invokes_step: [['ui_component', 'flow_step']],
  displayed_on: [['ui_component', 'screen']],
  triggered_by: [['ui_component', 'flow_step']],
};

export const isLinkAllowed = (type: LinkType, sourceKind: NodeKind, targetKind: NodeKind) =>
  (LINK_RULES[type] || []).some(([s, t]) => s === sourceKind && t === targetKind);

File: backend/app/ir/link_rules.py
from __future__ import annotations

from typing import Final

from .schema import LinkType, NodeKind


LINK_RULES: Final[dict[LinkType, set[tuple[NodeKind, NodeKind]]]] = {
    LinkType.REALIZES: {
        (NodeKind.CAPABILITY, NodeKind.GOAL),
        (NodeKind.FLOW, NodeKind.TASK),
    },
    LinkType.SUPPORTS: {
        (NodeKind.TASK, NodeKind.CAPABILITY),
        (NodeKind.FLOW_STEP, NodeKind.TASK),
    },
    LinkType.PERFORMED_BY: {
        (NodeKind.TASK, NodeKind.ACTOR),
        (NodeKind.FLOW_STEP, NodeKind.ACTOR),
    },
    LinkType.OWNS: {
        (NodeKind.ACTOR, NodeKind.BUSINESS_OBJECT),
    },
    LinkType.PRECEDES: {
        (NodeKind.FLOW_STEP, NodeKind.FLOW_STEP),
    },
    LinkType.BRANCHES_TO: {
        (NodeKind.FLOW_STEP, NodeKind.FLOW_STEP),
    },
    LinkType.GUARDS: {
        (NodeKind.RULE, NodeKind.FLOW_STEP),
        (NodeKind.RULE, NodeKind.STATE_TRANSITION),
    },
    LinkType.READS: {
        (NodeKind.FLOW_STEP, NodeKind.BUSINESS_OBJECT),
        (NodeKind.SCREEN, NodeKind.BUSINESS_OBJECT),
    },
    LinkType.WRITES: {
        (NodeKind.FLOW_STEP, NodeKind.BUSINESS_OBJECT),
        (NodeKind.SCREEN, NodeKind.BUSINESS_OBJECT),
    },
    LinkType.CHANGES_STATE: {
        (NodeKind.FLOW_STEP, NodeKind.STATE_TRANSITION),
    },
    LinkType.CONTAINS: {
        (NodeKind.GOAL, NodeKind.CAPABILITY),
        (NodeKind.CAPABILITY, NodeKind.CAPABILITY),
        (NodeKind.BUSINESS_OBJECT, NodeKind.FIELD),
        (NodeKind.BUSINESS_OBJECT, NodeKind.STATE_MACHINE),
        (NodeKind.STATE_MACHINE, NodeKind.OBJECT_STATE),
        (NodeKind.STATE_MACHINE, NodeKind.STATE_TRANSITION),
        (NodeKind.SCREEN, NodeKind.UI_COMPONENT),
        (NodeKind.UI_COMPONENT, NodeKind.UI_COMPONENT),
    },
    LinkType.ACCESSIBLE_BY: {
        (NodeKind.SCREEN, NodeKind.ACTOR),
    },
    LinkType.BINDS_FIELD: {
        (NodeKind.UI_COMPONENT, NodeKind.FIELD),
    },
    LinkType.INVOKES_STEP: {
        (NodeKind.UI_COMPONENT, NodeKind.FLOW_STEP),
    },
    LinkType.DEPENDS_ON: {
        (NodeKind.CAPABILITY, NodeKind.CAPABILITY),
        (NodeKind.TASK, NodeKind.TASK),
        (NodeKind.TASK, NodeKind.CAPABILITY),
        (NodeKind.FLOW_STEP, NodeKind.FLOW_STEP),
    },
    LinkType.DIAGNOSES: set(),
}


def is_link_allowed(link_type: LinkType | str, source_kind: NodeKind | str, target_kind: NodeKind | str) -> bool:
    link = link_type if isinstance(link_type, LinkType) else LinkType(link_type)
    source = source_kind if isinstance(source_kind, NodeKind) else NodeKind(source_kind)
    target = target_kind if isinstance(target_kind, NodeKind) else NodeKind(target_kind)
    return (source, target) in LINK_RULES.get(link, set())
